### ISSUE TITLE
TryGet Component?

### ISSUE THREAD

**Issue by Mason11987** [craftworkgames/MonoGame.Extended#474]
So I found myself in need of checking if an entity has a component, but I don't see any way of doing that.

I wrote a method, but I don't like it so I wouldn't do a pull request.  I suspect that the reason I need this is a code smell, but it really seems like a sensible feature.

Any thoughts on this feature?

My code I'm using right now in EntityManager is this:

 
       internal bool TryGetComponent<T>(Entity entity, out T component) where T : EntityComponent
        {
            component = null;
            Debug.Assert(entity != null);

            var componentType = GetComponentTypeFrom(typeof(T));

            Debug.Assert(entity != null);
            Debug.Assert(componentType != null);
            Debug.Assert(componentType.Index < _componentTypeEntitiesToComponents.Count);

            var components = _componentTypeEntitiesToComponents[componentType.Index];

            if (components == null)
                return false;

            EntityComponent entityComponent;
            components.TryGetValue(entity, out entityComponent);
            component = (T)entityComponent;
            return component != null;
        }

**Comment by stefanrbk** [comment]
The only thing I see missing from the code is the `EntityComponentType componentType` part is missing from the method arguments needed for the 12th line of code. I would have to look at the ECS more to see how the Component system even works to understand its use though. Do you have a use for this in mind already?

**Comment by Mason11987** [comment]
I have a system which interacts with several different kinds of entities.  That system should be applied on every entity with component A or B.  But if that component has C, I'd like to also use some of it's data, if it doesn't I don't want to.

I realize I could make a separate system that handles A, B, and C, and have this one handle ones with just A, and B, but that seems like it would complicate things a lot more than needed.  If I could just check if C exists, and if so use some of it's data, that'd be much easier, but as far as I can tell there's no real way to do that with the current ECS system.

**Comment by LithiumToast** [comment]
[This functionality already exists.](https://github.com/craftworkgames/MonoGame.Extended/blob/develop/Source/MonoGame.Extended.Entities/EntityManager.cs#L347)

You can simply just create a system with the following attribute aspect:

```
[Aspect(AspectType.All, typeof(A), typeof(B))]
```

Then in your system logic:

```
var component = entity.Get<C>();
// if component is null, the entity doesn't have that component attached
```

**Comment by Mason11987** [comment]
I was getting an exception thrown when I called Get<C> on a component that doesn't exist on an entity.  It was in a Debug.Assert though since I'm not using the compiled code, that may be the reason why.  I'll have to validate I have the latest code tonight and verify that still happens.

**Comment by Mason11987** [comment]
Okay, given EntityManager.cs

        internal EntityComponent GetComponent(Entity entity, EntityComponentType componentType)

Lines:

        var components = _componentTypeEntitiesToComponents[componentType.Index];
        Debug.Assert(components != null);

        EntityComponent component;
        components.TryGetValue(entity, out component);
        return component;

If no entities have the given component, the debug line will throw up an error.  I think a fix would be:

            var components = _componentTypeEntitiesToComponents[componentType.Index];

            EntityComponent component = null;
            components?.TryGetValue(entity, out component);
            return component;

I might make a pull request if I get a chance if no one else wants to address it first.

**Comment by LithiumToast** [comment]
If I remember correctly, I added a bunch of debug assertions cause I didn't trust the code that I imported from Artemis. That one assertion is irrelevant since the question mark on `components` is a `null` check anyways.

**Comment by Mason11987** [comment]
So this fix resolves a symptom of a possible issue, which is ultimately that _componentTypeEntitiesToComponents doesn't have an item for each type of component type index that exists.   It also contains null entries, when the code seems to expect it to not have nulls here (maybe elsewhere?)

I haven't dug into how _componentTypeEntititesToComponents is built or how scanning works at all, but I naively I would expect it to be an array of IEnumerables that start out empty, and never be null.  Is that reasonable?

**Comment by craftworkgames** [comment]
I presume this is fixed with #478
